content/articles/sharing-deployments-with-nixops.md

```diff
@@ -0,0 +1,198 @@
+$ nixops deploy
+test-keys> uploading EC2 key pair ‘charon-f29084f8-570c-11e5-83af-542696dce997-test-keys’...
+server...> creating EC2 instance (AMI ‘ami-0126a576’, type ‘t1.micro’, region ‘eu-west-1’)...
+server...> waiting for spot instance request ‘sir-01pq1vda’ to be fulfilled... [pending-evaluation] [fulfilled]
+server...> cancelling spot instance request ‘sir-01pq1vda’... [request-canceled-and-instance-running]
+server...> waiting for IP address... [pending] [running] 54.76.8.50 / 172.31.30.116
+server...> waiting for SSH......
+```
+
+
+### Trick 4: Sharing secrets in git
+
+Keeping AWS keys or any secrets in git unencrypted is a recipe for
+disaster. Bitcoin miners constantly scan Github for accidental AWS key
+uploads and will have spent thousands of your pounds before you can
+blink. The NixOps state file itself contains SSH root keys, exposing
+those keys is the same as having the machine owned with a zero day exploit.
+
+**Keeping either the statefile or the `.envrc` with your AWS keys
+in git unencrypted is grossly negligent.** I know this is common
+knowledge but I prefer to over-communicate on this issue.
+
+We're using a tool called
+[git-crypt](https://github.com/AGWA/git-crypt) to securely share
+secrets in a git repository. git-crypt creatively uses the
+`.gitattribute` file to encrypt certain files in the repository.
+
+First we initialize git-crypt:
+
+```console
+$ git-crypt init
+Generating key...
+$ git-crypt status
+not encrypted: .envrc
+not encrypted: localstate.nixops
+not encrypted: network.nix
+```
+
+Now we set up the `.gitattributes` file:
+
+```shell
+# .gitattributes
+localstate.nixops filter=git-crypt
+.envrc filter=git-crypt diff=git-crypt
+```
+
+Double-check the encryption works:
+
+```console
+$ git-crypt status
+    encrypted: .envrc
+not encrypted: .gitattributes
+    encrypted: localstate.nixops *** WARNING: diff=git-crypt attribute not set ***
+not encrypted: network.nix
+```
+
+Note that we didn't set `diff=git-crypt` on `localstate.nixops`
+because it is a binary SQLite file.
+
+
+### A safer .gitattributes
+
+Keeping encrypted files in many different places in the repository can
+cause accidental exposure due to human mistakes so we generally have a
+`secrets` directory for all the secret things (except for `.envrc`
+which needs to stay top-level).
+
+For example:
+
+```shell
+# .gitattributes
+secrets/* filter=git-crypt
+.envrc filter=git-crypt diff=git-crypt
+```
+
+### Pinning a specific nixos release
+
+As of Nix 1.9 one can refer to a specific release using HTTP, so we
+tend to pin our relase like so in the `.envrc` file:
+
+```shell
+# .envrc
+export NIX_PATH=nixpkgs=https://github.com/NixOS/nixpkgs-channels/archive/nixos-15.09.tar.gz:.
+```
+
+
+### Downsides of this scheme
+
+* Many commands like `deploy`, `backup`, etc modify the state file,
+requiring either a commit or the conscious choice to blast away state
+file changes with `git checkout -f` if you are sure it's safe to do
+so.
```

Comment:
deploy is fine you can just overwrite the state file

backup requires committing the change

This is the ugliest bit and the reason why I'd love to have a diff-able text format!